Client/Frontend/Browser/LoginsHelper.swift

```diff
@@ -180,48 +179,181 @@ class LoginsHelper: TabContentScript {
         snackBar?.addButton(update)
         tab?.addSnackbar(snackBar!)
     }
-
-    fileprivate func requestLogins(_ login: LoginData, requestId: String, frameInfo: WKFrameInfo) {
-        let currentHost = tab?.webView?.url?.host
-        let frameHost = frameInfo.securityOrigin.host
-        
-        profile.logins.getLoginsForProtectionSpace(login.protectionSpace).uponQueue(.main) { res in
-            var jsonObj = [String: Any]()
-            if let cursor = res.successValue {
-                log.debug("Found \(cursor.count) logins.")
-                jsonObj["requestId"] = requestId
-                jsonObj["name"] = "RemoteLogins:loginsFound"
-                jsonObj["logins"] = cursor.compactMap { loginData -> [String: String]? in
-                    if frameInfo.isMainFrame {
-                        return loginData?.toDict()
-                    }
-                    
-                    // The frame must belong to the same security origin
-                    if let currentHost = currentHost,
-                       !currentHost.isEmpty,
-                       currentHost == frameHost {
-                        // Prevent XSS on non main frame
-                        // If it is not the main frame, return username only, but no password!
-                        // Chromium does the same on iOS.
-                        // Firefox does NOT support third-party frames or iFrames.
-                        loginData?.update(password: "", username: loginData?.username ?? "")
-                        return loginData?.toDict()
+    
+    private func updateORSaveCredentials(for url: URL, script: [String: Any]) {
+        guard let scriptCredentials = passwordAPI.fetchFromScript(url, script: script),
+              let username = scriptCredentials.usernameValue,
+              scriptCredentials.usernameElement != nil,
+              let password = scriptCredentials.passwordValue,
+              scriptCredentials.passwordElement != nil else {
+            log.debug("Missing Credentials from script")
+            return
+        }
+        
+        if password.isEmpty {
+            log.debug("Empty Password")
+            return
+        }
+        
+        passwordAPI.getSavedLogins(for: url, formScheme: .typeHtml) { [weak self] logins in
+            guard let self = self else { return }
+            
+            for login in logins {
+                if (login.usernameValue ?? "").caseInsensitivelyEqual(to: username) {
```

Comment:
Do we want to fallback to empty string here?
Another option I see is to make this an early return check and if usernameValue is nil we `continue` the loop

Reply:
While selecting save or update the case you mentioned could be totally checked using a guard statement and continue the loop. (this case should never happen btw)
Fix at https://github.com/brave/brave-ios/pull/5013/commits/25f3d8d938c0ec02e77c7fa1a27e9ca9b275ce18